<launch>
    <!-- usage: roslaunch detection bringup_detection.launch planar:=false/true -->
    <arg name="planar" default="false" doc="The method for grasp detection, launch planar detection"/>
    <arg name="planar_grconv" default="false" doc="The method for grasp detection using grconv, launch planar grconv detection"/>
    <arg name="spatial" default="false" doc="The method for grasp detection, launch spatial detection."/>
    <arg name="gn1b" default="false" doc="The method for grasp detection, launch gn1b minimal detection."/>
    
    <!-- 相机内参 -->
    <param name="/detection/intrinsics/fx" type="double" value="607.239501953125" />
    <param name="/detection/intrinsics/fy" type="double" value="607.239501953125" />
    <param name="/detection/intrinsics/cx" type="double" value="327.332" />
    <param name="/detection/intrinsics/cy" type="double" value="243.108" />
    <!-- TODO fov的值待确认 -->
    <param name="/detection/intrinsics/fov" type="double" value="68" />

    <!-- 负责进行planar检测的节点 -->
    <node if="$(arg planar)" pkg="detection" type="node_detection.py" name="node_detection" output="screen" required="true" >
        <!-- 节点私有参数 -->
        <!-- 这个模型还行 -->
        <param name="model_path" value="ckpt/model_v2.tar"/>
        <param name="use_rgbd" value="true" />
        <param name="compensated" value="true"/>
        <param name="spatial" value="false" />
    </node>

    <node if="$(arg planar_grconv)" pkg="detection" type="node_detection_grconv.py" name="node_detection_grconv" output="screen" required="true" />

    <!-- 负责进行GN1B-MINIMAL检测的节点 -->
    <node if="$(arg gn1b)" pkg="detection" type="node_gn1b_minimal.py" name="node_detection_gn1b_minimal" output="screen" required="true" />

    <!-- realsense相机节点 -->
    <include file="$(find realsense2_camera)/launch/rs_camera.launch" >
        <arg name="device_type" value="d435"/>
        <arg name="align_depth" value="true"/>

        <arg name="depth_width" value="640"/>
        <arg name="depth_height" value="480"/>
        
        <arg name="color_width" value="640"/>
        <arg name="color_height" value="480"/>
        
        <arg name="enable_pointcloud" value="true"/>
    </include>

    <!-- 眼在手上的外参标定结果 -->
    <arg name="qx" value="-0.5146710160099053" />
    <arg name="qy" value="0.4963062370587139" />
    <arg name="qz" value="-0.5072768386750054" />
    <arg name="qw" value="0.4811071328507235" />
    <arg name="x" value="-0.24435441521426188" />
    <arg name="y" value="0.03398190911800965" />
    <arg name="z" value="0.1142620117698149" />
    
    <!-- TF发布：发布相机的外参：这里用的是眼在手上，所以发布相机和机械臂工具末端静态的变换关系 -->
    <!-- 旋转角度参数为：x y z qx qy qz qw -->
    <node name="camera_extrinsic_publisher" pkg="tf" type="static_transform_publisher" output="screen" respawn="true" args="$(arg x) $(arg y) $(arg z) $(arg qx) $(arg qy) $(arg qz) $(arg qw) grasping_frame camera_color_optical_frame 20" />
    
</launch>